covalent_ssh_plugin/ssh.py

```diff
@@ -241,35 +217,64 @@ async def _client_connect(self) -> Tuple[bool, asyncssh.SSHClientConnection]:
-            retries = 6 if self.retry_connect else 1
-            for _ in range(retries):
-                try:
-                    conn = await asyncssh.connect(
-                        self.hostname,
-                        username=self.username,
-                        client_keys=[self.ssh_key_file],
-                        known_hosts=None,
-                    )
 
-                    ssh_success = True
-                except (socket.gaierror, ValueError, TimeoutError, ConnectionRefusedError) as e:
-                    app_log.error(e)
+        if not os.path.exists(self.ssh_key_file):
+            message = f"SSH key file {self.ssh_key_file} does not exist."
+            app_log.error(message)
+            raise RuntimeError(message)
 
-                if conn is not None:
-                    break
+        try:
+            conn = await self._attempt_client_connect()
+            ssh_success = conn is not None
+        except (socket.gaierror, ValueError, TimeoutError) as e:
+            app_log.error(e)
 
-                await asyncio.sleep(5)
+        return ssh_success, conn
 
-            if conn is None and not self.run_local_on_ssh_fail:
-                raise RuntimeError("Could not connect to remote host.")
+    async def _attempt_client_connect(self) -> Optional[asyncssh.SSHClientConnection]:
+        """
+        Helper function that catches specific errors and retries connecting to the remote host.
 
-        else:
-            message = f"no SSH key file found at {self.ssh_key_file}. Cannot connect to host."
-            app_log.error(message)
-            raise RuntimeError(message)
+        Args:
+            max_attempts: Gives up after this many attempts.
 
-        return ssh_success, conn
+        Returns:
+            An `SSHClientConnection` object if successful, None otherwise.
+        """
+
+        # Retry connecting if any of these errors happen:
+        _retry_errs = (
+            ConnectionRefusedError,
+            OSError,  # e.g. Network unreachable
+        )
+
+        address = f"{self.username}@{self.hostname}"
+        attempt_max = self.max_connection_attempts
+
+        attempt = 0
+        while attempt < attempt_max:
+
+            try:
+                # Exit here if the connection is successful.
+                return await asyncssh.connect(
+                    self.hostname,
+                    username=self.username,
+                    client_keys=[self.ssh_key_file],
+                    known_hosts=None,
+                )
+            except _retry_errs as err:
+
+                if not self.retry_connect:
+                    app_log.error(f"{err} ({address} | retry disabled).")
+                    raise err
+
+                app_log.warning(f"{err} ({address} | retry {attempt+1}/{attempt_max})")
+                await asyncio.sleep(self.retry_wait_time)
+
+            finally:
+                attempt += 1
+
+        return None
```

Comment:
No need to return `None` explicitly here, it will be returned automatically.

Reply:
@kessler-frost But it's more readable with the explicit return, no? Happy to remove if you insist.

Reply:
That's a good point, but I'd say in almost every case there's usually a way to right the function in such a way where it can be avoided, but as always readable > concise. That said, it looks more natural to me now that you've added the `Failed to connect to client` comment 😄 so let's keep it this way.